mgr/dashboard: Enable compiler options used by Angular --strict flag

tspmelo opened the pull request:
Recently Angular added a new flag to the "ng new" command, that adds 5 new typescript compiler options.
They add more strict rules that help the code to be more readable and less prone to errors.
The PR can be found here: https://github.com/angular/angular-cli/pull/14905

Rules that will be applied in this PR:
**noImplicitAny** - https://basarat.gitbooks.io/typescript/docs/options/noImplicitAny.html
**noImplicitReturns** - https://lucidchart.com/techblog/2018/06/20/how-to-actually-improve-type-safety-with-the-typescript-strict-flags/
**noImplicitThis** - https://logicbig.com/tutorials/misc/typescript/no-implicit-this.html
**noFallthroughCasesInSwitch** - https://typescriptlang.org/docs/handbook/release-notes/typescript-1-8.html

**strictNullChecks** will be applied in a later PR.

<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

tspmelo commented:
jenkins test dashboard

tspmelo commented:
jenkins test make check

epuertat commented:
This is a real admirable work @tspmelo ! Did you use some script to automate it?

On the use of `CdObj<T>`, while it at least types the variable as a mapping, it appears >190 times as `CdObj<any>`, hence not performing any type check at all (beyond the mapping interface). I was wondering whether the increase in type safety is worthy the amount of changes required (as this massive clean-ups make backports harder to perform).

An alternative would be to use a Typescript utility type, like `Record<K, T>`. It would work as an alias for `CdObj<any>` with `Record<string, any>`. This way we could also support fast incremental typing: `Record<"field1" | "field2" | "field3", string>`?

tspmelo commented:
Thanks, most of it was manually or some search and replace.

The only reason I created the `CdObj` was to be able to use the mapping interface, since with the stricter options you can't used it anymore with `object`s.
I added the Generic to be able to be more strict in some situations, but it ended up being just few cases.

I tried other utilities types but none worked with using string as an index.
Didn't checked `Record` and that one seems to work :)

I just did a new search on ways to fix this problem and found that there is a compiler option that might help us, although reducing the strictness of the compiler.

I see this possible solutions:

1. Enable `suppressImplicitAnyIndexErrors `option.
This will allows us to access objects with a string index.
This will be like before and I could remove the use of `CdObj<T>` and remove all the `[k: string]: any;` I added to classes/interfaces.
This would reduce a lot of the changes made.
2. Replace all use of `CdObj<T>` with `Record<K, T>`
This will remove the need to import a new file while keeping all functionallity.
Still needs `[k: string]: any;` to be added to classes/interfaces.
3. Apply both 1 and 2.
We would apply option 1, but on situations where we have `CdObj<T>` that uses a non `any` type, we could use `Record<K, T>` to keep the strictness.

My vote is on the 3rd option.

epuertat commented:
@tspmelo, I'd take the option that:
1. Increases quality of existing code,
2. Enforces new code to meet the stricter rule,
3. Minimizes changes (for maintained back-portability).

I think your suggestion #3 fulfills all of those, so I'm happy with that.

tspmelo commented:
I have removed the use of `CdObj` and all new `[k: string]: any;`

tspmelo commented:
jenkins test dashboard

callithea commented:
jenkins test make check

tspmelo commented:
jenkins test dashboard

Devp00l commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/crushmap/crushmap.component.ts:
```diff
@@ -74,7 +75,8 @@ export class CrushmapComponent implements OnInit {
   }
 
   onNodeSelected(e: NodeEvent) {
-    const { name, type, status, ...remain } = this.metadataKeyMap[e.node.id];
```
Could we use `as number` instead of importing `lodash` just to make the compiler happy?
If yes I would prefer `as number` instead.

tspmelo replied:
Same as before, will revert it.

alfonsomthd commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-form/rbd-form.component.spec.ts:
```diff
@@ -59,11 +60,11 @@ describe('RbdFormComponent', () => {
+    let editAction: any;
+    let cloneAction: any;
+    let copyAction: any;
+    let rbdServiceGetSpy: any;
```
```suggestion
    let rbdServiceGetSpy: jasmine.Spy;
```

tspmelo replied:
Done

alfonsomthd commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cephfs/cephfs-directories/cephfs-directories.component.spec.ts:
```diff
@@ -35,15 +35,15 @@ describe('CephfsDirectoriesComponent', () => {
-  let maxBinaryValidator;
-  let originalDate;
-  let modal;
+  let lsDirSpy: any;
```
```suggestion
  let lsDirSpy: jasmine.Spy;
```

alfonsomthd commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/hosts.component.spec.ts:
```diff
@@ -20,7 +20,7 @@ import { HostsComponent } from './hosts.component';
   let component: HostsComponent;
   let fixture: ComponentFixture<HostsComponent>;
-  let hostListSpy;
+  let hostListSpy: any;
```
```suggestion
  let hostListSpy: jasmine.Spy;
```

alfonsomthd commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/osd/osd-details/osd-details.component.spec.ts:
```diff
@@ -20,7 +20,7 @@ describe('OsdDetailsComponent', () => {
   let debugElement: DebugElement;
   let osdService: OsdService;
-  let getDetailsSpy;
+  let getDetailsSpy: any;
```
```suggestion
  let getDetailsSpy: jasmine.Spy;
```

alfonsomthd commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/prometheus/silence-form/silence-form.component.spec.ts:
```diff
@@ -41,15 +41,15 @@ describe('SilenceFormComponent', () => {
   // Spies
-  let rulesSpy;
-  let ifPrometheusSpy;
+  let rulesSpy: any;
```
```suggestion
  let rulesSpy: jasmine.Spy;
```

alfonsomthd commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/prometheus/silence-form/silence-form.component.spec.ts:
```diff
@@ -174,7 +174,7 @@ describe('SilenceFormComponent', () => {
   describe('redirect not allowed users', () => {
     let prometheusPermissions: Permission;
-    let navigateSpy;
+    let navigateSpy: any;
```
```suggestion
    let navigateSpy: jasmine.Spy;
```

Devp00l commented on src/pybind/mgr/dashboard/frontend/src/app/shared/services/notification.service.ts:
```diff
@@ -171,7 +171,24 @@ export class NotificationService {
+        toastrCall = this.toastr.success;
+        break;
+      default:
+        return;
```
This doesn't seem code style related, it also makes the code less readable and produces a state that is never be used and if it were used (which should raise an error) no error would be thrown.

Devp00l commented on src/pybind/mgr/dashboard/frontend/src/app/shared/services/prometheus-alert-formatter.ts:
```diff
@@ -65,7 +65,8 @@ export class PrometheusAlertFormatter {
     };
-    return NotificationType[_.findKey(types, (type) => type.includes(status))];
+    const key = _.findKey(types, (type) => type.includes(status));
+    return _.get(NotificationType, key);
```
I wonder why this change is needed, as I don't see the difference and it's not more type save than before. The only difference is that `_.get` is taking a view more steps.

tspmelo replied:
With `noImplicitAny `enabled and `suppressImplicitAnyIndexErrors` disabled, TS is very strict with the type of the index used.
p.e. in a normal class you can't use any index to access its properties.

Since this is no longer the case I will revert.

tspmelo commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-configuration-form/rbd-configuration-form.component.ts:
```diff
@@ -77,10 +81,15 @@ export class RbdConfigurationFormComponent implements OnInit {
+      const control: any = this.form.get('configuration').get(config.name);
       const dirty = control.dirty;
 
-      if (this.initialData && this.initialData[config.name] === control.value) {
```
Hi, I think this caused an error before I enabled `suppressImplicitAnyIndexErrors`.
I will revert it since its no longer the case.

tspmelo commented on src/pybind/mgr/dashboard/frontend/e2e/ui/notification.e2e-spec.ts:
```diff
@@ -20,7 +20,7 @@ describe('Notification page', () => {
 
   it('should open notification sidebar', async () => {
-    await notification.waitStaleness(notification.getSidebar());
+    await notification.waitInvisibility(notification.getSidebar());
```
This was just to test the fix for the e2e, already created a separate PR, will remove it.

tspmelo commented on src/pybind/mgr/dashboard/frontend/src/app/shared/components/pwd-expiration-notification/pwd-expiration-notification.component.ts:
```diff
@@ -37,11 +37,9 @@ export class PwdExpirationNotificationComponent implements OnInit {
-    }
+    const current = new Date();
+    const expiration = new Date(pwdExpirationDate * 1000);
+    return Math.floor((expiration.valueOf() - current.valueOf()) / (1000 * 3600 * 24));
```
If I revert this I will need to add `return undefined;` at the end of the function.
I though this was a better fix.

Devp00l replied:
If `pwdExpirationData` is not set you will call `Date` with `NaN` this would return 'NaN' at the end. I'm not sure if it will break something but it has much more potential to do so.

Devp00l replied:
I tested it locally doesn't seem to break anything.

tspmelo commented on src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.ts:
```diff
@@ -304,7 +304,7 @@ export class TableComponent implements AfterContentChecked, OnInit, OnChanges, O
     return columns
       .reduce(
-        (result, value, index) =>
+        (result, value, index: number) =>
```
Looks like it does.

tspmelo commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.ts:
```diff
@@ -411,20 +411,24 @@ export class PoolFormComponent implements OnInit {
       return pgs / size;
     }
+
+    return undefined;
```
This is due to `noImplicitReturns`.
I'm OK with disabling it, if requested.

But I think the code is actually more readable and you will never forget to return something or miss that its not returning when reviewing.

Also, I prefer the current code, it might have more lines but its more readable.

Devp00l replied:
We should vote for it during the next stand up :)
I'm for disabling it, as how a method/function works should be known.

LenzGr replied:
> But I think the code is actually more readable and you will never forget to return something or miss that its not returning when reviewing.

FWIW, I too am in favor of better readability. Keep in mind that we should write code that others could quickly get familiar with in order to lower the barrier for new contributors (and for reviewers, FWIW).

I suggest to keep the check for implicit returns - this also improves the readability and avoids ambiguity.

p-se replied:
I tend to be more strict as well. I like the `noImplicitReturns` rule and I think (and hope) it will eventually lead to code that looks like it was written in a strictly typed language. That is, when declared to return a number, will return a number. Taking the example, it'd look like this:

```typescript
private replicatedPgCalc(pgs: number): number {
  const sizeControl = this.form.get('size');
  const size = sizeControl.value;
  if (sizeControl.valid && size > 0) {
    return pgs / size;
  }
  return 0; // or: -1   <---
}
```

As alternative, the method/function can throw an exception which can be caught by the caller.

This looks way more natural to me than writing `return undefined`, which by the way, feels unnatural to me for two reasons:

1. The method declared a number to be returned but returned nothing
2. `undefined` is the return value for attributes of objects which aren't there or method/function arguments that haven't been provided. I'd prefer to use `null` if necessary, but would even try to avoid that by either returning a variable of the declared type or raising an exception.

Either way, **the way to improve in that regard is to keep `noImplicitReturns`** and eventually write code that doesn't feel so unnatural. But as a first step I wouldn't require the code to be changed this way in this PR, as this is the current behavior the code and the PR is really big enough, even without changing its behavior.

Devp00l replied:
Any changes can me made in a follow up PR to get it in quickly.

Devp00l commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-configuration-form/rbd-configuration-form.component.ts:
```diff
@@ -1,6 +1,8 @@
 import { Component, EventEmitter, Input, OnInit, Output } from '@angular/core';
 import { FormControl, Validators } from '@angular/forms';
 
+import * as _ from 'lodash';
```
You forgot to remove the import.
